pallets/foreign-investments/src/lib.rs

```diff
@@ -437,20 +355,320 @@ pub mod pallet {
+								)?;
+								SwapIds::<T>::insert(who, investment_id, id);
+							}
+						}
+					}
+				}
+			}
+
+			Ok(())
+		}
+
+		fn decrease_foreign_investment(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+			foreign_amount: T::Balance,
+			foreign_currency: T::CurrencyId,
+			pool_currency: T::CurrencyId,
+		) -> DispatchResult {
+			let pool_amount = T::CurrencyConverter::stable_to_stable(
+				pool_currency,
+				foreign_currency,
+				foreign_amount,
+			)?;
+
+			let id = SwapIds::<T>::get(who, investment_id).ok_or(Error::<T>::SwapOrderNotFound)?;
+			let swap = T::TokenSwaps::get_order_details(id).ok_or(Error::<T>::SwapOrderNotFound)?;
+
+			if swap.currency_out == foreign_currency {
+				match swap.amount.cmp(&pool_amount) {
+					Ordering::Less => T::TokenSwaps::update_order(
+						who.clone(),
+						id,
+						swap.amount.ensure_sub(pool_amount)?,
+						T::BalanceRatio::one(),
+					)?,
+					Ordering::Equal => T::TokenSwaps::cancel_order(id.clone())?,
+					Ordering::Greater => {
+						T::TokenSwaps::cancel_order(id.clone())?;
+
+						let foreign_swap_amount = T::CurrencyConverter::stable_to_stable(
+							foreign_currency,
+							pool_currency,
+							swap.amount,
+						)?;
+
+						let decrease_investment_pool_amount =
+							pool_amount.ensure_sub(swap.amount)?;
+
+						T::Investment::update_investment(
+							who,
+							investment_id,
+							T::Investment::investment(who, investment_id)?
+								.ensure_sub(decrease_investment_pool_amount)?,
+						)?;
+
+						let id = T::TokenSwaps::place_order(
+							who.clone(),
+							foreign_currency,
+							pool_currency,
+							foreign_amount.ensure_sub(foreign_swap_amount)?,
+							T::BalanceRatio::one(),
+						)?;
+						SwapIds::<T>::insert(who, investment_id, id);
+					}
+				}
+			} else {
+				T::Investment::update_investment(
+					who,
+					investment_id,
+					T::Investment::investment(who, investment_id)?.ensure_sub(pool_amount)?,
+				)?;
+
+				T::TokenSwaps::update_order(
+					who.clone(),
+					id,
+					swap.amount.ensure_add(foreign_amount)?,
+					T::BalanceRatio::one(),
+				)?;
+			}
+
+			Ok(())
+		}
+
+		fn increase_foreign_redemption(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+			amount: T::Balance,
+			_payout_currency: T::CurrencyId,
+		) -> DispatchResult {
+			T::Investment::update_redemption(
+				who,
+				investment_id,
+				T::Investment::redemption(who, investment_id)?.ensure_add(amount)?,
+			)
+		}
+
+		fn decrease_foreign_redemption(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+			amount: T::Balance,
+			_payout_currency: T::CurrencyId,
+		) -> Result<(T::Balance, T::Balance), DispatchError> {
+			T::Investment::update_redemption(
+				who,
+				investment_id,
+				T::Investment::redemption(who, investment_id)?.ensure_sub(amount)?,
+			)?;
+
+			// For William: Why I should return here the amounts? Should not be done after
+			// collecting the redemption?
+
+			Ok(todo!())
+		}
```

Comment:
Wondering about this

Reply:
Gotcha, we just return the trache tokens that becomes "free" again after decreasing the redemption